presto-verifier/src/main/java/com/facebook/presto/verifier/framework/InsertExtendedVerification.java

```diff
@@ -0,0 +1,177 @@
+import com.facebook.presto.verifier.checksum.ChecksumValidator;
+import com.facebook.presto.verifier.prestoaction.QueryActions;
+import com.facebook.presto.verifier.prestoaction.SqlExceptionClassifier;
+import com.facebook.presto.verifier.resolver.FailureResolverManager;
+import com.facebook.presto.verifier.rewrite.QueryRewriter;
+import com.facebook.presto.verifier.source.SnapshotQueryConsumer;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ListeningExecutorService;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+
+import static com.facebook.presto.verifier.framework.DataMatchResult.DataType.PARTITION_DATA;
+import static com.facebook.presto.verifier.framework.DataMatchResult.MatchType.ROW_COUNT_MISMATCH;
+import static com.facebook.presto.verifier.framework.DataVerificationUtil.getColumns;
+import static com.facebook.presto.verifier.framework.DataVerificationUtil.match;
+import static com.facebook.presto.verifier.framework.QueryStage.CONTROL_PARTITION_CHECKSUM;
+import static com.facebook.presto.verifier.framework.QueryStage.TEST_PARTITION_CHECKSUM;
+import static com.facebook.presto.verifier.framework.VerifierUtil.callAndConsume;
+
+public class InsertExtendedVerification
+        extends DataVerification
+{
+    public InsertExtendedVerification(
+            QueryActions queryActions,
+            SourceQuery sourceQuery,
+            QueryRewriter queryRewriter,
+            DeterminismAnalyzer determinismAnalyzer,
+            FailureResolverManager failureResolverManager,
+            SqlExceptionClassifier exceptionClassifier,
+            VerificationContext verificationContext,
+            VerifierConfig verifierConfig,
+            TypeManager typeManager,
+            ChecksumValidator checksumValidator,
+            ListeningExecutorService executor,
+            SnapshotQueryConsumer snapshotQueryConsumer,
+            Map<String, SnapshotQuery> snapshotQueries)
+    {
+        super(
+                queryActions,
+                sourceQuery,
+                queryRewriter,
+                determinismAnalyzer,
+                failureResolverManager,
+                exceptionClassifier,
+                verificationContext,
+                verifierConfig,
+                typeManager,
+                checksumValidator,
+                executor,
+                snapshotQueryConsumer,
+                snapshotQueries);
+    }
+
+    @Override
+    public DataMatchResult verify(
+            QueryObjectBundle control,
+            QueryObjectBundle test,
+            Optional<QueryResult<Void>> controlQueryResult,
+            Optional<QueryResult<Void>> testQueryResult,
+            ChecksumQueryContext controlChecksumQueryContext,
+            ChecksumQueryContext testChecksumQueryContext)
```

Comment:
You might notice that the base class DataVerification.verify() supports several workflow branches
```
if (isControlEnabled()) {
    if (saveSnapshot) {
    }
}
else if (QUERY_BANK_MODE.equals(runningMode)) {
   if (snapshotQuery != null) {
   }
   else {
    }
}
```

The big thing I see here we need to respect all the workflow branches, and make sure no one of them breaks. Ex., some workflow branches disable control or disable checksum. So the user won't set a valid control cluster to run. So we cannot assume there's a control run before this point.

I agree with you we inherit the DataVerification and run this InsertExtendedVerification only for Insert and CAST